File: ios/hls/HlsCacheStore.swift
import CommonCrypto
import Foundation

final class HlsCacheStore {
  private let maxBytes: Int = 5_368_709_120
  private let ttlSeconds: TimeInterval = 7 * 24 * 60 * 60
  private let cacheDir: URL
  private let indexUrl: URL
  private var index: [String: HlsCacheEntry] = [:]
  private let queue = DispatchQueue(label: "hls-cache-store")
  private var pendingSave: DispatchWorkItem?

  init() {
    let base = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask)[0]
    cacheDir = base.appendingPathComponent("hls-cache", isDirectory: true)
    indexUrl = cacheDir.appendingPathComponent("index.json")
    loadIndex()
    ensureDir()
  }

  func get(url: String) -> Data? {
    guard let entry = queue.sync(execute: { index[url] }) else { return nil }
    if isExpired(entry) {
      queue.async { [weak self] in
        self?.remove(url: url)
      }
      return nil
    }

    let fileUrl = cacheDir.appendingPathComponent(entry.fileName)
    guard let data = try? Data(contentsOf: fileUrl) else {
      queue.async { [weak self] in
        self?.remove(url: url)
      }
      return nil
    }

    queue.async { [weak self] in
      guard var current = self?.index[url] else { return }
      current.lastAccess = Date().timeIntervalSince1970
      self?.index[url] = current
      self?.scheduleSave()
    }

    return data
  }

  func put(url: String, data: Data, streamKey: String?) {
    queue.async {
      self.ensureDir()
      let name = self.sha256(url) + ".seg"
      let fileUrl = self.cacheDir.appendingPathComponent(name)
      do {
        try data.write(to: fileUrl, options: .atomic)
        let entry = HlsCacheEntry(
          url: url,
          fileName: name,
          size: data.count,
          streamKey: streamKey,
          createdAt: Date().timeIntervalSince1970,
          lastAccess: Date().timeIntervalSince1970
        )
        self.index[url] = entry
        self.evictIfNeeded()
        self.scheduleSave()
      } catch {
        return
      }
    }
  }

  func has(url: String) -> Bool {
    queue.sync {
      guard let entry = index[url] else { return false }
      if isExpired(entry) {
        remove(url: url)
        return false
      }
      let fileUrl = cacheDir.appendingPathComponent(entry.fileName)
      return FileManager.default.fileExists(atPath: fileUrl.path)
    }
  }

  func getFilePath(url: String) -> URL? {
    queue.sync {
      guard let entry = index[url] else { return nil }
      if isExpired(entry) {
        remove(url: url)
        return nil
      }
      let fileUrl = cacheDir.appendingPathComponent(entry.fileName)
      guard FileManager.default.fileExists(atPath: fileUrl.path) else {
        remove(url: url)
        return nil
      }
      var current = entry
      current.lastAccess = Date().timeIntervalSince1970
      index[url] = current
      scheduleSave()
      return fileUrl
    }
  }

  func getCacheStats() -> [String: Any] {
    queue.sync {
      evictExpired()
      let totalSize = index.values.reduce(0) { $0 + $1.size }
      return [
        "totalSize": totalSize,
        "fileCount": index.count,
        "maxSize": maxBytes
      ]
    }
  }

  func getCacheStats(streamKey: String) -> [String: Any] {
    queue.sync {
      evictExpired()
      let totalSize = index.values.reduce(0) { $0 + $1.size }
      let streamEntries = index.values.filter { $0.streamKey == streamKey }
      return [
        "totalSize": totalSize,
        "fileCount": index.count,
        "maxSize": maxBytes,
        "streamSize": streamEntries.reduce(0) { $0 + $1.size },
        "streamFileCount": streamEntries.count
      ]
    }
  }

  func clearAll() {
    queue.async {
      for entry in self.index.values {
        let fileUrl = self.cacheDir.appendingPathComponent(entry.fileName)
        try? FileManager.default.removeItem(at: fileUrl)
      }
      self.index.removeAll()
      self.saveIndex()
    }
  }

  private func ensureDir() {
    if !FileManager.default.fileExists(atPath: cacheDir.path) {
      try? FileManager.default.createDirectory(at: cacheDir, withIntermediateDirectories: true)
    }
  }

  private func evictIfNeeded() {
    evictExpired()
    var total = index.values.reduce(0) { $0 + $1.size }
    if total <= maxBytes { return }
    let entries = index.values.sorted { $0.lastAccess < $1.lastAccess }
    for entry in entries {
      if total <= maxBytes { break }
      remove(url: entry.url)
      total = index.values.reduce(0) { $0 + $1.size }
    }
  }

  private func evictExpired() {
    let now = Date().timeIntervalSince1970
    let entries = Array(index.values)
    for entry in entries where now - entry.createdAt > ttlSeconds {
      remove(url: entry.url)
    }
  }

  private func remove(url: String) {
    guard let entry = index[url] else { return }
    let fileUrl = cacheDir.appendingPathComponent(entry.fileName)
    try? FileManager.default.removeItem(at: fileUrl)
    index.removeValue(forKey: url)
    scheduleSave()
  }

  private func isExpired(_ entry: HlsCacheEntry) -> Bool {
    let now = Date().timeIntervalSince1970
    return now - entry.createdAt > ttlSeconds
  }

  private func loadIndex() {
    guard let data = try? Data(contentsOf: indexUrl) else { return }
    if let decoded = try? JSONDecoder().decode([String: HlsCacheEntry].self, from: data) {
      index = decoded
    }
  }

  private func scheduleSave() {
    pendingSave?.cancel()
    let work = DispatchWorkItem { [weak self] in
      self?.saveIndex()
    }
    pendingSave = work
    queue.asyncAfter(deadline: .now() + 5, execute: work)
  }

  private func saveIndex() {
    if let data = try? JSONEncoder().encode(index) {
      try? data.write(to: indexUrl, options: .atomic)
    }
  }

  private func sha256(_ input: String) -> String {
    let data = Data(input.utf8)
    var hash = [UInt8](repeating: 0, count: Int(CC_SHA256_DIGEST_LENGTH))
    data.withUnsafeBytes {
      _ = CC_SHA256($0.baseAddress, CC_LONG(data.count), &hash)
    }
    return hash.map { String(format: "%02x", $0) }.joined()
  }
}

struct HlsCacheEntry: Codable {
  let url: String
  let fileName: String
  let size: Int
  let streamKey: String?
  let createdAt: TimeInterval
  var lastAccess: TimeInterval
}

File: android/src/main/java/com/yupi/hls/HlsCacheStore.kt
package com.yupi.hls

import android.content.Context
import org.json.JSONObject
import java.io.File
import java.security.MessageDigest
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledFuture
import java.util.concurrent.TimeUnit

class HlsCacheStore(context: Context) {
    private val maxBytes: Long = 5L * 1024L * 1024L * 1024L
    private val ttlMs: Long = 7L * 24L * 60L * 60L * 1000L
    private val cacheDir = File(context.cacheDir, "hls-cache")
    private val indexFile = File(cacheDir, "index.json")
    private val index = ConcurrentHashMap<String, HlsCacheEntry>()
    private val saveExecutor = Executors.newSingleThreadScheduledExecutor()
    private var pendingSave: ScheduledFuture<*>? = null

    init {
        if (!cacheDir.exists()) cacheDir.mkdirs()
        loadIndex()
    }

    fun has(url: String): Boolean {
        val entry = index[url] ?: return false
        if (isExpired(entry)) {
            remove(url)
            return false
        }
        return File(cacheDir, entry.fileName).exists()
    }

    fun getFilePath(url: String): File? {
        val entry = index[url] ?: return null
        if (isExpired(entry)) {
            remove(url)
            return null
        }
        val file = File(cacheDir, entry.fileName)
        if (!file.exists()) {
            remove(url)
            return null
        }
        entry.lastAccess = System.currentTimeMillis()
        scheduleSave()
        return file
    }

    fun get(url: String): ByteArray? {
        val entry = index[url] ?: return null
        if (isExpired(entry)) {
            remove(url)
            return null
        }
        val file = File(cacheDir, entry.fileName)
        if (!file.exists()) {
            remove(url)
            return null
        }
        entry.lastAccess = System.currentTimeMillis()
        scheduleSave()
        return file.readBytes()
    }

    fun put(url: String, data: ByteArray, streamKey: String?) {
        if (!cacheDir.exists()) cacheDir.mkdirs()
        val name = sha256(url) + ".seg"
        val file = File(cacheDir, name)
        file.writeBytes(data)
        index[url] = HlsCacheEntry(url, name, data.size.toLong(), streamKey, System.currentTimeMillis(), System.currentTimeMillis())
        evictIfNeeded()
        scheduleSave()
    }

    fun getCacheStats(): Map<String, Any> {
        evictExpired()
        return mapOf(
            "totalSize" to index.values.sumOf { it.size },
            "fileCount" to index.size,
            "maxSize" to maxBytes
        )
    }

    fun getStreamCacheStats(streamKey: String): Map<String, Any> {
        evictExpired()
        val streamEntries = index.values.filter { it.streamKey == streamKey }
        return mapOf(
            "totalSize" to index.values.sumOf { it.size },
            "fileCount" to index.size,
            "maxSize" to maxBytes,
            "streamSize" to streamEntries.sumOf { it.size },
            "streamFileCount" to streamEntries.size
        )
    }

    fun clearAll() {
        index.values.forEach { entry ->
            val file = File(cacheDir, entry.fileName)
            if (file.exists()) file.delete()
        }
        index.clear()
        saveIndex()
    }

    private fun evictIfNeeded() {
        evictExpired()
        var total = index.values.sumOf { it.size }
        if (total <= maxBytes) return
        val entries = index.values.sortedBy { it.lastAccess }
        for (entry in entries) {
            if (total <= maxBytes) break
            total -= entry.size
            remove(entry.url)
        }
    }

    private fun evictExpired() {
        val now = System.currentTimeMillis()
        index.values.filter { now - it.createdAt > ttlMs }.forEach { remove(it.url) }
    }

    private fun remove(url: String) {
        val entry = index[url] ?: return
        val file = File(cacheDir, entry.fileName)
        if (file.exists()) file.delete()
        index.remove(url)
        scheduleSave()
    }

    private fun isExpired(entry: HlsCacheEntry): Boolean {
        return System.currentTimeMillis() - entry.createdAt > ttlMs
    }

    private fun loadIndex() {
        if (!indexFile.exists()) return
        val text = indexFile.readText()
        if (text.isBlank()) return
        val json = JSONObject(text)
        json.keys().forEach { key ->
            val obj = json.getJSONObject(key)
            val entry = HlsCacheEntry(
                url = key,
                fileName = obj.getString("fileName"),
                size = obj.getLong("size"),
                streamKey = if (obj.has("streamKey") && !obj.isNull("streamKey")) obj.getString("streamKey") else null,
                createdAt = obj.getLong("createdAt"),
                lastAccess = obj.getLong("lastAccess")
            )
            index[key] = entry
        }
    }

    private fun scheduleSave() {
        pendingSave?.cancel(false)
        pendingSave = saveExecutor.schedule({
            saveIndex()
        }, 5, TimeUnit.SECONDS)
    }

    @Synchronized
    private fun saveIndex() {
        val json = JSONObject()
        index.forEach { (url, entry) ->
            val obj = JSONObject()
            obj.put("fileName", entry.fileName)
            obj.put("size", entry.size)
            obj.put("streamKey", entry.streamKey)
            obj.put("createdAt", entry.createdAt)
            obj.put("lastAccess", entry.lastAccess)
            json.put(url, obj)
        }
        indexFile.writeText(json.toString())
    }

    private fun sha256(input: String): String {
        val digest = MessageDigest.getInstance("SHA-256").digest(input.toByteArray())
        return digest.joinToString("") { "%02x".format(it) }
    }
}

data class HlsCacheEntry(
    val url: String,
    val fileName: String,
    val size: Long,
    val streamKey: String?,
    val createdAt: Long,
    var lastAccess: Long
)
